src/util/range.h

```diff
@@ -265,6 +265,87 @@ struct concat_iteratort
   second_iteratort second_begin;
 };
 
+/// Zip two ranges to make a range of pairs.
+/// On increment, both iterators are incremented.
+/// Ends when the two ranges reach their ends, raise an exception if one ends
+/// before the other.
+template <typename first_iteratort, typename second_iteratort>
+struct zip_iteratort
+{
+public:
+  using difference_type = typename first_iteratort::difference_type;
+  using value_type = std::pair<
+    typename first_iteratort::value_type,
+    typename second_iteratort::value_type>;
+  using pointer = value_type *;
+  using reference = value_type &;
+  using iterator_category = std::forward_iterator_tag;
+
+  bool operator==(const zip_iteratort &other) const
+  {
+    return first_begin == other.first_begin && first_end == other.first_end &&
+           second_begin == other.second_begin && second_end == other.second_end;
+  }
+
+  bool operator!=(const zip_iteratort &other) const
+  {
+    return !(*this == other);
+  }
+
+  /// Preincrement operator
+  zip_iteratort &operator++()
+  {
+    ++first_begin;
+    ++second_begin;
+    INVARIANT(
+      (first_begin == first_end) == (second_begin == second_end),
+      "Zipped ranges should have the same size");
+    current = first_begin != first_end && second_begin != second_end
+                ? std::make_shared<value_type>(*first_begin, *second_begin)
+                : nullptr;
+    return *this;
+  }
+
+  /// Postincrement operator
+  const zip_iteratort operator++(int)
+  {
+    zip_iteratort tmp(first_begin, first_end, second_begin, second_end);
+    this->operator++();
+    return tmp;
+  }
+
+  reference operator*() const
+  {
+    return *current;
+  }
+
+  pointer operator->() const
+  {
+    return current.get();
+  }
+
+  zip_iteratort(
+    first_iteratort first_begin,
+    first_iteratort first_end,
+    second_iteratort second_begin,
+    second_iteratort second_end)
+    : first_begin(std::move(first_begin)),
+      first_end(std::move(first_end)),
+      second_begin(std::move(second_begin)),
+      second_end(std::move(second_end))
+  {
+    if(first_begin != first_end && second_begin != second_end)
```

Comment:
This will silently swallow the case where one range is empty but not the other, which per docs should lead to an exception